ariadne/cardano/src/Ariadne/Wallet/Cardano/Kernel/Consistency.hs

```diff
@@ -0,0 +1,57 @@
+-- | This module contains functions that are needed
+-- to check consistency of wallets and key storage.
+
+module Ariadne.Wallet.Cardano.Kernel.Consistency
+       ( getUnknownKeys
+       , getUnknownWallets
+       ) where
+
+import Data.List ((\\))
+import qualified Data.Map as Map
+
+import Ariadne.Wallet.Cardano.Kernel (PassiveWallet, getWalletSnapshot)
+import Ariadne.Wallet.Cardano.Kernel.DB.AcidState (dbHdWallets)
+import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet
+  (HdRoot, HdRootId, WalletName, hdRootId, hdRootName)
+import qualified Ariadne.Wallet.Cardano.Kernel.DB.HdWallet.Read as HDRead
+import qualified Ariadne.Wallet.Cardano.Kernel.DB.Util.IxSet as IxSet
+import Ariadne.Wallet.Cardano.Kernel.Keystore (Keystore)
+import qualified Ariadne.Wallet.Cardano.Kernel.Keystore as Keystore
+
+-- | Checks keystore for unknown keys and return list of them.
+getUnknownKeys :: MonadIO m => PassiveWallet -> Keystore -> m [HdRootId]
+getUnknownKeys pwallet keystore = do
+  (keystoreRootIds,walletRootIds,_) <- getRootIdsFromWalletAndKeyfile pwallet keystore
+  return $ (keystoreRootIds \\ walletRootIds)
+
+-- | Returns list of wallets without corresponding key in keystore.
+getUnknownWallets
+  :: MonadIO m => PassiveWallet -> Keystore -> m ([WalletName], [HdRootId])
+getUnknownWallets pwallet keystore = do
+  (keystoreRootIds,walletRootIds, wallets) <- getRootIdsFromWalletAndKeyfile pwallet keystore
+  let rootIDsWithMissedSecretKeys = (walletRootIds \\ keystoreRootIds)
+  let eitherWalletsWithMissedSecretKeys =
+        mapM ((flip IxSet.getTheOnly) wallets) $ rootIDsWithMissedSecretKeys
+  let walletNamesWithMissedSecretKeys = case eitherWalletsWithMissedSecretKeys of
+        Right wallet -> view hdRootName <$> wallet
+        Left  (brokenRootId, brokenWallets) ->
+                 error $ "There are more (or less) then one wallet with the same HdRootId: \n" <>
+                         "Broken RootId: " <> (show brokenRootId) <> "\n" <>
+                         "Names of wallets with this id: " <>
+                         (show $ view hdRootName <$> IxSet.toList brokenWallets)
```

Comment:
It's better to use printing function based on `Buildable` (e. g. one from the `fmt` package).